[terra-date-input-field] error message and focus indicator updates

KV106606Viswanath opened the pull request:
### Summary
1. Added focus to the first input field when the toggle button has been clicked. Now the screen reader reads the error message immediately when the button has been clicked.
2. Added black dashed border for input field when it has focus.

**What was changed:**


**Why it was changed:**



### Testing
<!-- Demonstrate that these changes are stable. How have these changes been verified? -->

This change was tested using:

- [ ] WDIO
- [ ] Jest
- [ ] Visual testing (please attach a screenshot or recording)
- [ ] Other (please describe below)
- [ ] No tests are needed

### Reviews

In addition to engineering reviews, this PR needs:
<!-- Please include the appropriate "Required" & "Ready" labels (ex. "UX Review Required" and "UX Review Ready").  -->
- [ ] UX review
- [ ] Accessibility review
- [ ] Functional review

### Additional Details
<!-- List anything else that is relevant to this issue. Additional information will help us better understand your changes and speed up the review process. -->

**This PR resolves:**

UXPLATFORM-10106 & UXPLATFORM-10107

---

Thank you for contributing to Terra.
@cerner/terra

sycombs commented on packages/terra-date-input/CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@
 
 * Added
   * Added hot keys functionality to support keyboard shortcut patterns.
+  * Added black dashed border when input gets focussed.
```
```suggestion
  * Added black dashed border when input gets focused.
```

KV106606Viswanath replied:
Updated

sycombs commented on packages/terra-date-input/src/DateInput.jsx:
```diff
@@ -606,7 +606,7 @@ class DateInput extends React.Component {
         <select
           {...this.props.monthAttributes}
-          aria-label={label}
+          aria-label={this.props.isInvalid ? 'Error message' : label}
```
This should probably be internationalized.

KV106606Viswanath replied:
The string "Error message" is not required so it has been removed.

sycombs commented on packages/terra-date-input/src/DateInput.jsx:
```diff
@@ -677,7 +677,7 @@ class DateInput extends React.Component {
         {...numberAttributes}
         refCallback={(inputRef) => { this.dayRef = inputRef; }}
-        label={label}
+        label={this.props.isInvalid ? `${'Error message'}, ${label}` : label}
```
Same as above, 'Error message' should probably be internationalized.

KV106606Viswanath replied:
The string "Error message" is not required so it has been removed.

sycombs commented on packages/terra-date-input/src/DateInput.jsx:
```diff
@@ -719,7 +719,7 @@ class DateInput extends React.Component {
         {...numberAttributes}
         refCallback={(inputRef) => { this.yearRef = inputRef; }}
-        label={this.props.intl.formatMessage({ id: 'Terra.date.input.yearLabel' })}
+        label={this.props.isInvalid ? `${'Error message'}, ${this.props.intl.formatMessage({ id: 'Terra.date.input.yearLabel' })}` : `${this.props.intl.formatMessage({ id: 'Terra.date.input.yearLabel' })}`}
```
Same as above, 'Error message' should probably be internationalized.

KV106606Viswanath replied:
The string "Error message" is not required so it has been removed.

sycombs commented on packages/terra-date-input/tests/wdio/__snapshots__/reference/orion-fusion-theme/en/chrome_medium/date-input-field-spec/clear_month.png:
Is the blue highlight from the original border style intended to still be here? I think we've removed it for other components.

KV106606Viswanath replied:
Removed.

KV106606Viswanath commented on packages/terra-date-input/src/DateInputField.jsx:
```diff
@@ -136,6 +140,7 @@ const defaultProps = {
   showOptional: false,
   value: '',
   yearAttributes: {},
+  refCallback: undefined,
```
Removed.

KV106606Viswanath commented on packages/terra-date-input/CHANGELOG.md:
```diff
@@ -6,6 +6,8 @@
 * Added
   * Added hot keys functionality to support keyboard shortcut patterns.
+  * Added black dashed border when input gets focused.
+  * Added screen reader support to associate with error message.
```
Updated.

rbsree commented:
Regarding this PR, the points they have mentioned:
- Provided a black border for the input field that has focus. 
- When the toggle button is triggered, then the focus moves to the first input field. 
are now fixed, expect the error message that appears is not exposed by screen reader when the focus moves to the first input field when the toggle button is triggered.

The screen does read out "Invalid data/entry" when the focus is on the first input field, this is because of the aria-invalid="true" added to it and the error message is not associated with the input field using "aria-describedby" attribute.

KV106606Viswanath commented:
> Regarding this PR, the points they have mentioned:
> 
> * Provided a black border for the input field that has focus.
> * When the toggle button is triggered, then the focus moves to the first input field.
>   are now fixed, except the error message that appears is not exposed by the screen reader when the focus moves to the first input field when the toggle button is triggered.
> 
> The screen does read out "Invalid data/entry" when the focus is on the first input field, this is because of the aria-invalid="true" added to it and the error message is not associated with the input field using the "aria-describedby" attribute.

Updated.

sugan2416 commented:
@KV106606Viswanath Can we add a wdio test for the error message display ?

rbsree commented:
+1
As, the issues mentioned in my previous comments are fixed now, the error message that appears when toggle button is triggered is exposed by screen reader when focused.

supreethmr commented on packages/terra-date-input/src/DateInput.jsx:
```diff
@@ -637,6 +640,7 @@ class DateInput extends React.Component {
           <option key={this.props.intl.formatMessage({ id: 'Terra.date.input.december' })} value="12">{this.props.intl.formatMessage({ id: 'Terra.date.input.december' })}</option>
         </select>
         <VisuallyHiddenText id={this.hotKeyInstructionId} text={intl.formatMessage({ id: 'Terra.date.input.hotKey' })} />
+        <VisuallyHiddenText id={this.errorMessageId} text={`, ${this.props.error}`} />
```
we do not need visually hidden text as we already have a visual text for error message as shown in below screenshot.

<img width="1250" alt="Screenshot 2024-01-18 at 7 33 08 PM" src="https://github.com/cerner/terra-framework/assets/12869809/c0a76c48-7e8e-45fa-bb78-62e0d0d3558d">

KV106606Viswanath replied:
Updated

KV106606Viswanath commented:
> @KV106606Viswanath Can we add a wdio test for the error message display ?

Added

supreethmr commented on packages/terra-date-input/src/DateInput.jsx:
```diff
@@ -102,6 +102,8 @@ const propTypes = {
    * Custom input attributes to apply to the year input
    */
   yearAttributes: PropTypes.object,
+  errorMessageId: PropTypes.string,
+  ariadescribedby: PropTypes.string,
```
we need to add Prop description for above two props !!

KV106606Viswanath replied:
These are removed as per latest changes.

supreethmr commented on packages/terra-date-input/src/DateInput.jsx:
```diff
@@ -116,6 +118,8 @@ const defaultProps = {
   useExternalFormatMask: false,
   isA11yControlled: false,
+  errorMessageId: null,
+  ariadescribedby: null,
```
Not able to find `ariadescribedby` prop in DateInput, where is it being used ?

KV106606Viswanath replied:
This is removed as per latest changes.

supreethmr commented on packages/terra-date-input/src/DateInputField.jsx:
```diff
@@ -113,6 +113,11 @@ const propTypes = {
+   * Callback ref to access date input containing DOM element.
+   */
+  refCallback: PropTypes.func,
+  errorMessageId: PropTypes.string,
```
prop description is missing

KV106606Viswanath replied:
This is removed as per latest changes.

supreethmr commented on packages/terra-date-input/src/DateInputField.jsx:
```diff
@@ -136,6 +141,7 @@ const defaultProps = {
   showOptional: false,
   value: '',
   yearAttributes: {},
+  errorMessageId: null,
```
Same here, where is `errorMessageId` being used

KV106606Viswanath replied:
This is removed as per latest changes.

supreethmr commented on packages/terra-time-input/src/_AccessibleInput.jsx:
```diff
@@ -81,8 +84,7 @@ function AccessibleInput(props) {
         aria-label={value ? undefined : label}
-        aria-describedby={descriptionId}
-
+        aria-describedby={isInvalid && ariadescribedby ? `${descriptionId} ${ariadescribedby}` : `${descriptionId}`}
```
```suggestion
        aria-describedby={ariadescribedby ? `${ariadescribedby} ${descriptionId}` : `${descriptionId}`}
```

KV106606Viswanath replied:
Updated.

supreethmr commented on packages/terra-date-input/src/DateInput.jsx:
```diff
@@ -620,7 +623,7 @@ class DateInput extends React.Component {
           aria-invalid={this.props.isInvalid}
           aria-required={this.props.required}
-          aria-describedby={this.hotKeyInstructionId}
+          aria-describedby={`${this.hotKeyInstructionId} ${this.props.errorMessageId}`}
```
```suggestion
          aria-describedby={ (customProps['aria-describedby']) ?`${this.hotKeyInstructionId} ${customProps['aria-describedby']}` : this.hotKeyInstructionId}
```
this allow consumer passed aria-describedBy to be passed as well

KV106606Viswanath replied:
Updated.

supreethmr commented on packages/terra-date-input/src/DateInput.jsx:
```diff
@@ -102,6 +102,10 @@ const propTypes = {
+  /**
+   * Custom props to identify the error message node
+   */
+  customProps: PropTypes.Object,
```
we don't need declaration for customProps

KV106606Viswanath replied:
Removed.

supreethmr commented on packages/terra-date-input/src/DateInput.jsx:
```diff
@@ -696,6 +702,7 @@ class DateInput extends React.Component {
         showIsInvalid
         isIncomplete={this.props.isIncomplete}
         required={this.props.required}
+        ariadescribedby={this.props.customProps['aria-describedby']}
```
nit : can we get this extract on top instead of accessing it through object

KV106606Viswanath replied:
Updated.

saket2403 commented on packages/terra-time-input/src/_AccessibleInput.jsx:
```diff
@@ -24,11 +24,16 @@ const propTypes = {
+  /**
+   * String that labels the error message for screen readers.
+   */
+  ariadescribedby: PropTypes.string,
```
nit:
```suggestion
  ariaDescribedBy: PropTypes.string,
```

KV106606Viswanath replied:
Updated.

saket2403 commented on packages/terra-date-input/CHANGELOG.md:
```diff
@@ -6,6 +6,8 @@
 
 * Added
   * Added hot keys functionality to support keyboard shortcut patterns.
+  * Added black dashed border when input gets focused.
+  * Added screen reader support for error messages.
```
Pls move this entry to unreleased section

KV106606Viswanath replied:
Updated.

saket2403 commented on packages/terra-framework-docs/CHANGELOG.md:
```diff
@@ -2,6 +2,9 @@
 ## Unreleased
 
+* Added
+  * Added screen reader support to associate with error message.
```
Change this entry to relevant 
```suggestion
  * Updated `date-input` examples to allow screen reader response for error message.
```

KV106606Viswanath replied:
Updated.
